### PR TITLE
Console: read-only memory-engine panel, Brain re-listed, selection-scope decision recorded

### PR DESCRIPTION
**Pull request by YellowSnnowmann**
## Summary

P2 + P3 of the memory-engine surfacing plan, in one PR. **Stacked on #1273** (needs its `MemorySpec.healthy` field; review from commit `4268475b`). Merge order: #1248 → #1273 → this.

### P2 — the operator can finally SEE the engine
- **Brain returns to the nav** (parked by #302, re-listed deliberately: an operator choosing between memory engines needs somewhere to look). The parked-view comments are updated to match.
- **Engine panel on the Brain view**: driver, mode, negotiated capability chips, and the boot probe rendered as a green / red / neutral state (`reachable` / `unreachable — check the endpoint and credential` / `not probed`). A hosted engine serving exactly the mandatory three families says so out loud — the spec's operator rights assume richer families that live only engine-side (the G13 surfacing half; the backend already refuses over-claims at bind, and no oc port can currently reach an unsupported family, so there is no typed-unsupported error to add yet).
- **Memory engine card in Settings**: the same facts as read-only InfoRows beside the Connection card.
- Both surfaces render nothing on the `store` default and on hosts predating the `/spec` memory field.

### P3 — the decision, recorded where it binds
- `docs/spec/runtime/memory-engine.md` gains the selection-scope record: **infra-operator only, instance-wide, env at boot** — no per-company/per-agent selection, no API or Console setter, and why it deliberately does not follow the per-company `[inference]` model. The two new UI surfaces carry no setter for exactly this reason, and their doc-comments say so.

## Verification
- `npm run typecheck` and `npm run build` in `frontend/` — clean
- No Rust changes; no e2e spec touched (the existing `memory-engine-badge` testid is preserved; the new panel and card add `memory-engine-panel` / `settings-memory-engine` testids for future specs)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added selectable memory-engine configuration for hosted and embedded deployments.
  - Restored the Memory view as “Brain” in the sidebar.
  - Added read-only engine details, capabilities, backend mode, and boot health indicators.
  - Added startup health checks for configured memory engines.

- **Bug Fixes**
  - Improved settings lifecycle handling to prevent stale controls during refreshes.

- **Documentation**
  - Added memory-engine configuration, deployment, switching, and troubleshooting guidance.

- **Tests**
  - Expanded memory-provider conformance and health-status coverage.

### CODE REVIEW

**Review by oxoxDev — CHANGES_REQUESTED** [review]
The panel is clean — three-states `Option<bool>` correctly, leaks nothing, genuinely read-only, and the P3 reasoning holds up against the code it cites. Two things block.

**First, credit where I owe it:** the staleness concern I raised on #1273 is already addressed here, and better than I expected. `MemoryView.tsx:563-565` reads "reachable **at boot**" / "unreachable **at boot** — check the endpoint and credential" / "not probed", and Settings hedges structurally instead — the InfoRow is labelled "Boot probe" and the card says "read at boot", so its bare "reachable" is scoped by its own label. That clears the bar I said was owed. The only residual is that neither surface says *when* boot was, so "reachable at boot" on a 30-day-old pod is true but unhelpful, and the green dot still reads as "live" faster than the text beside it does. A timestamp is a `/spec` change so I'm not asking for it here; a less status-light-coded treatment for the dot might be worth a thought.

**Major — `backend === "null"` renders as an ordinary healthy engine, on the one view with a write button.**

`MemoryView.tsx:526` and `SettingsView.tsx:350` gate on `backend === "store"` only. But `MemoryBackend::Null` (`select.rs:165-166`, "writes accepted and discarded, reads empty") binds through `open_provider`, gets probed, and reports `healthy: true`. So an operator sees `Engine: <driver>` · `Mode: null` · **green dot, "reachable at boot"** — beside a working-looking "New memory" button whose writes vanish.

The host already treats this as the alarming case. `select.rs:661-669` emits a dedicated warn: *"OPENCOMPANY_MEMORY=null is bound: memory writes are accepted and discarded… Nothing this company is told will be remembered."* — with the stated reason that every read returning empty is indistinguishable from a company that simply hasn't learned anything yet. Moving that from the pod log into the operator's eye is exactly what this panel is for, and the panel's own docstring says it exists "so an operator can see what a hosted engine does NOT support before a cycle discovers it". Rendering the loudest case as a normal green row inverts that.

One conditional fixes it — `backend === "null"` gets warning treatment carrying the discard sentence. I'd suppress or warn on the Add button too, but the copy is the minimum.

**Major — the Brain e2e was left skipped, and its comment now says the opposite of the truth.**

`frontend/test/e2e/wiring.spec.ts:71-78`:

> *"Parked by issue #302: the console no longer lists Brain… this stays here verbatim to be **un-skipped the day the surface is relisted**, rather than deleted and rewritten from memory."*

This is that day, and the `test.skip` and comment are untouched. So the re-listed view and both new surfaces ship with zero automated coverage, a repo comment is now factually false, and the body's "the parked-view comments are updated to match" is incomplete.

Compounding it: `EnginePanel` isn't exported, so the `Option<bool>` → three-state mapping — a pure function, trivially testable, and the exact thing an operator's trust rests on — can't be unit-tested either. `npm test` is 148 files / 1417 tests; this repo doesn't ship UI logic untested. Either un-skip the wiring spec (it should pass — `Console E2E` runs a real host binary), or export `EnginePanel` and unit-test the three states plus the `store`/`null`/old-host gates.

**Non-blocking**

- `app-shell.tsx:166` — the `HIDDEN_VIEWS` comment still lists `memory` among the parked views. The two you updated at `:156` and `:177` are good; this one now disagrees with them ten lines apart.
- `MemoryView.tsx:584` — "Richer families (trees, graph, **sources**)". Both the Rust (`store/memory/mod.rs:147`) and this PR's own doc (`memory-engine.md:69-70`) say "no summary tree, no graph and no **taint** tier"; `sources` appears nowhere. Also "are not served over a hosted engine" over-generalizes from *this* engine to all hosted ones — the condition is `mandatoryOnly`, a statement about the bound engine.
- `MemoryView.tsx:509` — `MANDATORY_FAMILIES` is a hardcoded frontend copy of a backend contract asserted at `routes.rs:925-927` and `store/memory/test.rs:578-580`, with no coupling test. A backend rename silently stops the note rendering with nothing going red — same class as the classifier-string-drift trap.

**Nits:** `healthy: null` renders a transparent dot with no label (`MemoryView.tsx:554-566`) — currently unreachable thanks to `skip_serializing_if`, so defensive only, but an `else` would make the render total. The P3 section at `memory-engine.md:329-339` near-duplicates the runbook paragraph three lines below it. And the `memory-engine-badge` pill at `:231-247` now says the same thing as the panel directly beneath it — retaining the testid for the e2e is fair, but the visible pill may be redundant.

**What I verified clean:** no leakage — `MemorySpec` carries exactly `backend`, `driver_id`, `capabilities`, `healthy`, with **no `reason` string**, so there's no channel for an error message to smuggle an endpoint, and the panel renders a strict subset. All three `Option<bool>` states map correctly and `undefined` cannot render as red (checks are strict `=== false`). Genuinely read-only on both surfaces — no setter, no mutation path, no disabled-but-present control implying a future one.

**On re-listing Brain — safe.** I looked up #302 and it was a pure MVP-IA declutter, which went out of its way to say *"Removing the nav item is not removing the capability"* and *"No backend route, store or test is deleted as part of this — nav only."* Not parked for being incomplete or misleading, and the premise has lapsed anyway since NAV now carries Ledgers, Approvals, Workflows, Workspace and Company. One note: this reverses a maintainer IA decision without a sign-off cited, so probably worth a ping rather than a block.

**Stack drift worth catching:** this doesn't carry #1273's `69dac733` ("privatize the probe handle, pin the merge commit, unclaim the union") — the commit that addresses my #1273 review. Merge #1273 first and rebase, or this PR's merge re-introduces the pre-review Rust.

CI is clean — everything green except `Rust (openhuman, tinycortex)` still running at normal duration, and all the Rust lanes are inherited #1273 content anyway since this PR has zero Rust. I reproduced locally: all three typecheck projects clean, `npx vitest run` 148 files / 1417 passed. Note `Console E2E` passes *without* exercising Brain, per the skipped spec. tinysweeper's approval is pinned one commit stale and is an embedding-only pass; CodeRabbit hasn't posted at this head. Its walkthrough bullet about "improved settings lifecycle handling" is misattributed — there's no lifecycle change, it's describing the new card's own `live`/`setEngine(undefined)` cleanup guard.

File: src/store/select.rs [hunk 1]
```diff
@@ -256,6 +315,13 @@ pub struct MemoryDescriptor {
     /// The capability families the bound driver negotiated, so an operator can
     /// see what the engine does *not* support before a cycle finds out.
     pub capabilities: Vec<String>,
+    /// Whether the boot-time reachability probe answered `Ready`.
+    ///
+    /// `None` means the engine was never probed — the engine-overlay path,
+    /// which has no provider seam to ask, or a boot path that skipped
+    /// [`MemoryOverlay::refresh_health`]. `Some(false)` is a bound engine
+    /// whose probe failed: still bound, loudly warned, visible on `/spec`.
+    pub healthy: Option<bool>,
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/store/select.rs]: Fixed across the stack: the two Rust field docs in 75dfe5d9 (on #1273, where those files belong — this branch carries it via merge) and the TS contract comment in 5dc3eb55. All three now say usable = Ready or Degraded; only Down is false.

File: Cargo.toml [hunk 2]
```diff
@@ -578,6 +578,15 @@ whisper-rs-sys = { git = "https://github.com/tinyhumansai/whisper-rs-sys.git", b
...
 tempfile = "3"
+
+# The upstream driver-conformance suite (`assert_provider`, `retains_writes`),
+# run in `src/store/memory/upstream_conformance_test.rs` against the providers
+# `open_driver` constructs. Unconditional on purpose — dev-dependencies cannot
+# be optional — and cheap on purpose: the crate is contract-only
+# (tinymemory-api + async-trait + serde_json + anyhow), deliberately barred
+# from reaching any engine, so it adds no engine weight to any lane's graph.
+# Same byte-identical-path rule as the `tinymemory*` entries above.
+tinymemory-conformance = { path = "vendor/openhuman/vendor/tinymemory/conformance" }
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply Cargo.toml:589]: False positive from an uninitialized submodule: vendor/openhuman is a git submodule, so the path is empty in a clone without `--recursive`. The crate exists at vendor/openhuman/vendor/tinymemory/conformance (contract-only: its manifest depends on tinymemory-api + tokio/serde/axum for the doubles), resolves in Cargo.lock, and the gated CI lane compiled and ran its 8 conformance tests on this PR (`Rust (openhuman, tinycortex)`, green). No change.

File: src/bin/opencompany.rs [hunk 3]
```diff
@@ -1486,7 +1486,16 @@ async fn async_main() -> Result<()> {
...
+            if let Some(mut overlay) = opencompany::store::open_memory_overlay(&storage_settings)? {
+                // One bounded reachability probe, so a dead endpoint or a
+                // revoked key shows on `/spec` at boot instead of surfacing as
+                // a mid-cycle failure days later. Advisory: it warns and
+                // records, never refuses — config errors already refused
+                // above, and a transient vendor outage must not crash-loop
+                // the tenant.
+                overlay
+                    .refresh_health(std::time::Duration::from_secs(5))
+                    .await;
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/bin/opencompany.rs:1498]: Declining the extraction: the ordering is enforced by the borrow checker rather than by a test — `refresh_health` takes `&mut overlay` and `with_memory_overlay(overlay)` moves it, so a probe placed after the handoff cannot compile on that binding. The supplied tests pin the refresh semantics themselves (Ready/Degraded/Down/no-seam), and a boot-path assertion through async_main would need a full env harness for one line of glue. The serve-probe placement is also under discussion on #1273 (post-bind-pre-listener wording) — if it moves post-listener a helper will fall out naturally there.

File: src/store/memory/upstream_conformance_test.rs [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+
+    async fn cg_remember(
+        State(sets): State<Datasets>,
+        form: axum::extract::Multipart,
+    ) -> Json<Value> {
+        let (body, dataset, filename) = multipart_parts(form).await;
+        let mut sets = sets.lock().expect("store lock");
+        sets.entry(dataset).or_default().insert(filename, body);
+        Json(json!({ "status": "ok" }))
+    }
```
[the remark this answers is not in the document]
Comment (YellowSnnowmann) [reply src/store/memory/upstream_conformance_test.rs:582]: Declining locally, filing upstream: these doubles are deliberate near-verbatim copies of the tinymemory maintainers' own conformance doubles (24 of 25 functions whitespace-identical — verified in the #1273 review), and that provenance is what makes them trustworthy as vendor-shape oracles. Diverging our copy's multipart handling would fork the oracle. Both points (the while-let Ok swallow and the empty-filename key collapse) belong in the upstream double, alongside the cg_recall shape bug we already carry a divergence note for — filing them together on tinymemory. The local total-loss guard (retains_writes before assert_provider) still catches the full-collapse case.

**Comment by YellowSnnowmann** [comment]
@oxoxDev Both majors are addressed in `bb80e2c9`, plus most of the non-blockers:

**null renders as the alarming case now.** `enginePanelMode()` classifies `backend === "null"` as `discard`: the dot goes amber regardless of the probe (green would be a lie beside a write button), and a destructive Alert carries the host's own discard sentence — accepts-and-discards, reads-come-back-empty, indistinguishable-from-never-learned, with the env var to unset. Settings' card description carries the same sentence. I stopped short of disabling the Add button — the copy is doing the warning's job and a dead button invites "is this broken?" — but happy to add suppression if you'd rather.

**The Brain e2e runs again — and on the default lane, not just live-brain.** Un-skipping in place would have left it gated: `wiring.spec.ts` is file-level `test.skip(!LIVE_BRAIN)` because it proves the mocked inference chain, while this flow exercises `…/memory` and needs no inference at all. So the test moved verbatim to its own `brain-memory.spec.ts` (history noted in its header), ungated — the default `Console E2E` job now executes it instead of skipping it. Verified locally against the real host binary: passes in 542ms. The panel logic is also now unit-testable and tested: `enginePanelMode` / `probeLabel` (total over the domain, including a literal `null`) / `isMandatoryOnly` exported pure, 9 new tests, suite at 150 files / 1427.

Non-blockers: `app-shell` enumeration fixed; families copy now says summary tree/graph/**taint** and scopes the claim to *this* engine; `MANDATORY_FAMILIES` carries the drift-trap pointer at the two Rust assertion sites (a generated contract is the real fix; out of scope here); the render is total (no blank label on any `healthy` value); the P3/runbook near-dup deduped. Kept: the pill above the panel (testid is load-bearing for the e2e, and the pill is the glanceable summary of the panel's detail).

On re-listing Brain reversing #302's IA decision: flagged to the product owner rather than assumed — the re-listing was their direct request, but your "ping rather than block" is noted in the PR body now too.

Stack drift: already healed before your review posted — the branch carries #1273's `69dac733` (and its `75dfe5d9` doc fix) via merge; merge order stays #1273 → this.

**Review by oxoxDev — APPROVED** [review]
Approving. Both blockers are genuinely fixed, and I verified them by rendering the panel and by reading the CI log rather than by inspection.

**`null` backend — fixed.** The logic routes through an exported `enginePanelMode()`, and I rendered `EnginePanel` directly across all four states:

| input | dot | discard alert |
|---|---|---|
| `backend:"null", healthy:true` | `bg-status-blocked` (amber) | present |
| `backend:"null", healthy:false` | `bg-status-blocked` | present |
| `backend:"remote", healthy:true` | `bg-status-done` (green) | absent |
| `store` / old host | renders `""` | — |

Amber is a real token (`--status-blocked`, `index.css:232`) used the same way in `status-pill` and `host-switcher`, and the `store`/old-host gates still return null — confirmed by render, not by reading the branch. The copy carries both the meaning and the lever: *"accepts and discards every write — nothing this company is told will be remembered … Adding memories here does nothing until the infra operator unsets `OPENCOMPANY_MEMORY=null`."* Settings got the same treatment on its `CardDescription`. Leaving the Add button live is fine given the Alert sits directly beneath the header row that contains it and says adding does nothing.

**Brain e2e — fixed, and better than what I asked for.** Rather than un-skipping in place you moved it to `brain-memory.spec.ts`, and the reasoning holds: `wiring.spec.ts:7` carries a file-level `test.skip(!LIVE_BRAIN, …)`, so un-skipping in place would have left it gated on `PW_LIVE_BRAIN`. The new spec has no such gate — and it actually ran, from the `Console E2E` log at head:

```
✓ 26 test/e2e/brain-memory.spec.ts:28:1 › operator adds a Brain memory that persists across reload and can be deleted (854ms)
23 skipped / 192 passed (5.4m)
```

Green in the live-brain lane too. And the nine new unit tests are load-bearing — I mutation-tested them and all four mutations were caught: dropping the `null → "discard"` branch, inverting `probeLabel`, reducing `isMandatoryOnly` to a length check, and emptying the `probeLabel` fallback.

Non-blocking items: the `HIDDEN_VIEWS` comment is fixed, and the families copy is fixed on both halves — "summary tree, graph, taint" now matches `MemoryTaint` and the doc's own line 70, and "this engine serves only the floor" replaces the over-general phrasing. `MANDATORY_FAMILIES` stays uncoupled but now names the two Rust assertion sites in a comment; the unit test uses the constant as its own oracle so a backend rename still drifts silently — reasonable as recorded debt. The `healthy: null` label is total and explicitly tested, and the doc duplication is now a one-line back-reference.

**Two small things for whenever you're next in here, neither worth another round:**

- The blocker-1 remedy itself has no committed test — only the *decision* is covered. I deleted the entire `{mode === "discard" && <Alert …>}` block and the full suite still passed 1426/1426; nothing references `memory-engine-discard`, `memory-engine-panel` or `settings-memory-engine`. A `renderToStaticMarkup` assertion closes it in about five lines (RTL isn't installed, so `react-dom/server` is the path — I confirmed it works).
- `SettingsView` duplicates the classification inline (`backend === "store"` gate, `const discarding = engine.backend === "null"`) instead of importing `enginePanelMode`. Two surfaces, one contract, no shared source — it drifts if a future backend adds another discard-shaped mode.

The badge pill still duplicates the panel beneath it and still renders for `null` in neutral styling, which is cosmetic but is the element physically adjacent to the Add button while the Alert sits below. And the null panel still reads "reachable at boot" beside an amber dot — literally true, you flagged the tension in-comment, and the Alert disambiguates immediately.

Stack drift is resolved: merge-base is exactly `75dfe5d9` and this carries `69dac733`, so no out-of-order-merge risk. Gates clean locally — all three typecheck projects exit 0, `npx vitest run` 149 files / 1426 tests, build fine. CI fully green at head; `Rust (openhuman, tinycortex)` resolved to success at 30m33s — it was queue-slow, not broken, with five sibling runs of the same lane in flight simultaneously.

One caveat on the bot row, as usual: tinysweeper's approval is at a two-commits-stale SHA on an embedding-only pass (`0 in / 0 out`, sub-checks reporting "Reviewed 0 files"), and CodeRabbit's check reads `pass` on "Review rate limited". Its stale `CHANGES_REQUESTED` at `ac213933` was never re-submitted, though it did explicitly withdraw all four inline findings at head — all of which were on #1273's Rust rather than this increment.
